walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,140 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // The total (common) free time we have to conduct the meeting.
+    ArrayList<TimeRange> free_time = new ArrayList<TimeRange>();
+    free_time.addAll(Arrays.asList(TimeRange.WHOLE_DAY));
+
+    int meeting_duration = (int) request.getDuration();
+
+    // If duration of the meeting is greater than a whole day, then return an empty list.
+    if (meeting_duration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // If there are no attendees then return the whole day (ie the free time we have till now).
+    if (request.getAttendees().isEmpty()) {
+      return free_time;
+    }
+
+    /**
+     * Go through every event. If an event has attendees of the requested meeting, then check if the
+     * time overlaps with our free time. If yes then remove that conflicted time slot from our free
+     * time.
+     */
+    for (Event event : events) {
+      if (!Collections.disjoint(new ArrayList<>(request.getAttendees()), event.getAttendees())) {
+        int conflicted_slot_index = containsEvent(free_time, event.getWhen());
+        if (conflicted_slot_index != -1) {
+          while(conflicted_slot_index < free_time.size()) {
+            //loop through free time slots till the event ends and remove it from our free time.
+            TimeRange free_slot = free_time.get(conflicted_slot_index);
+            free_time =
+                getNewFreeTime(free_slot, event.getWhen(), free_time, conflicted_slot_index);
+            conflicted_slot_index++;
+            //break if the event ends in one of the free slots
+            if (event.getWhen().end() < free_slot.end()) break;
+          }
+        }
+      }
+    }
+
+    // If any of the TimeRange in our freetime list has a
+    // duration less than the requested meeting duration then remove that slot.
+    free_time = checkDuration(free_time, meeting_duration);
+
+    return free_time;
+  }
+
+  // This method checks if an event overlaps with any of our free time slots.
+  // It returns the conflicted slot's
+  private int containsEvent(ArrayList<TimeRange> free_time, TimeRange event_timerange) {
```

Comment:
`containsEvent` is a misleading name. The comments suggests, the function should be called `overlapsWithEvent`.

Reply:
Hm yes "overlapsWithEvent" sounds more accurate.